### PR TITLE
Set up autopropose service environment

### PR DESCRIPTION
**Pull request by adaszko**
Set up logging, tests, pipenv environment, and mypy.  Next step: call gRPC API via pyrchain.

### CODE REVIEW

File: autopropose.py [hunk 1]
```diff
@@ -1,71 +1,138 @@
...
+from ecdsa import SigningKey
+from ecdsa.curves import SECP256k1
+
+
+logging.basicConfig(level=logging.INFO, format='%(levelname)s %(message)s')
 LOGGER = logging.getLogger()
 
 
-def parse_conf(contents):
+def generate_key_pair_hex() -> Tuple[str, str]:
```
Comment (woky) [inline autopropose.py]: Why don't you just depend on pyrchain?

File: autopropose.py [hunk 2]
```diff
@@ -1,71 +1,138 @@
...
+        '--phlo-price', '1',
+        '--private-key', private_key,
+        contract,
+    ]
+
+    try:
+        LOGGER.info("Deploying %s", contract)
+        subprocess.check_output(deploy_command)
+    except subprocess.CalledProcessError as e:
+        LOGGER.error(e.cmd)
```
Comment (woky) [inline autopropose.py]: How about the following?
```
        LOGGER.error("Command: " + e.cmd)
        LOGGER.exception('deploy')
        LOGGER.error("Exit code: " + e.returncode)
        LOGGER.error("Stdout:\n" + e.stdout)
        LOGGER.error("Stderr:\n" + e.stderr)
```

File: autopropose.py [hunk 3]
```diff
@@ -1,71 +1,138 @@
...
+        LOGGER.exception('propose')
+        LOGGER.error(e.returncode)
+        LOGGER.error(e.stdout)
+        LOGGER.error(e.stderr)
+
+
+@contextlib.contextmanager
+def must_take_at_least(seconds: int) -> Iterator[None]:
+    started_at = int(time.monotonic())
+    try:
```
Comment (woky) [inline autopropose.py]: What does this construction mean?
```
try:
    yield
finally:
    # ... actual code ...
```

File: autopropose.py [hunk 4]
```diff
@@ -1,71 +1,138 @@
...
-        if env_var in os.environ:
-            deploy_env[env_var] = os.environ[env_var]
+def main() -> int:
+    config = parse_conf_file('/var/lib/rnode-static/autopropose.conf')
 
-    LOGGER.info('Deploying contract %s every %d seconds', deploy_env.get('contract'), deploy_no_sooner_than_every)
+    contract = config.get('contract')
+    if contract is None or contract == '' or contract == 'none':
+        LOGGER.info('No contract configured.  Going to sleep...')
+        while True:
```
Comment (woky) [inline autopropose.py]: I don't get this. If no contract is configured, then `main()` effectively loops infinitely? It should still propose every `period`.

File: run-setup [hunk 5]
```diff
@@ -103,36 +103,27 @@ systemctl reload nginx
...
-	if [[ $# -ne 1 ]]; then
-	    echo "restart-unit-if-not-masked: $#: wrong number of arguments"
-	    return 1
-	fi
-	local unit; unit=$1; shift
+if [[ ${PROFILE:-} = sandboxnet ]]; then
+    apt install --yes --no-install-recommends python3-pip
+    python3 -m pip install --user pipenv
+    /root/.local/bin/pipenv sync
+    install --compare --mode=644 autopropose.py.service --no-target-directory /usr/lib/systemd/system/autopropose.service
```
Comment (woky) [inline run-setup]: OK, I get that you prefer long option names. But please, at least make it easier for me to read by wrapping the lines at 80th column. I (and others) divide screen estate into several windows.

File: bootstrap.sandboxnet [hunk 6]
```diff
@@ -6,7 +6,7 @@ REPO_DIR=/opt/rchain-testnet-node
 GIT_CRYPT_KEY_FILE=/root/rchain-sre-git-crypt-key
 SETUP_SCRIPT_URL=https://raw.githubusercontent.com/rchain/rchain-testnet-node/master/update-run-setup
 PROFILE=sandboxnet
-REPO_BRANCH=dev
+REPO_BRANCH=deploy-propose
```
Comment (woky) [inline bootstrap.sandboxnet]: Is this supposed to be changed to `dev` before merge? Please also squash other adequate groups of your commits (e.g. `XXX` with something else).

File: autopropose.py [hunk 7]
```diff
@@ -1,71 +1,138 @@
...
+
+    try:
+        LOGGER.info("Deploying %s", contract)
+        subprocess.check_output(deploy_command)
+    except subprocess.CalledProcessError as e:
+        LOGGER.error(e.cmd)
+        LOGGER.exception('deploy')
+        LOGGER.error(e.returncode)
+        LOGGER.error(e.stdout)
+        LOGGER.error(e.stderr)
```
Comment (woky) [inline autopropose.py]: How about the following?
```
        LOGGER.error("Command: " + e.cmd)
        LOGGER.exception('deploy')
        LOGGER.error("Exit code: " + e.returncode)
        LOGGER.error("Stdout:\n" + e.stdout)
        LOGGER.error("Stderr:\n" + e.stderr)
```

File: autopropose.py [hunk 8]
```diff
@@ -1,71 +1,138 @@
...
+def must_take_at_least(seconds: int) -> Iterator[None]:
+    started_at = int(time.monotonic())
+    try:
+        yield
+    finally:
+        finished_at = int(time.monotonic())
+        took = finished_at - started_at
+        if took < seconds:
+            wait_duration = seconds - took
+            time.sleep(wait_duration)
```
Comment (woky) [inline autopropose.py]: What does this construction mean?
```
try:
    yield
finally:
    # ... actual code ...
```

File: autopropose.py [hunk 9]
```diff
@@ -1,71 +1,138 @@
...
-            deploy_env[env_var] = os.environ[env_var]
+def main() -> int:
+    config = parse_conf_file('/var/lib/rnode-static/autopropose.conf')
 
-    LOGGER.info('Deploying contract %s every %d seconds', deploy_env.get('contract'), deploy_no_sooner_than_every)
+    contract = config.get('contract')
+    if contract is None or contract == '' or contract == 'none':
+        LOGGER.info('No contract configured.  Going to sleep...')
+        while True:
+            time.sleep(1)
```
Comment (woky) [inline autopropose.py]: I don't get this. If no contract is configured, then `main()` effectively loops infinitely? It should still propose every `period`.

File: autopropose.py (hunk 1 above)
Comment (adaszko) [reply autopropose.py]: That's the next step.

File: bootstrap.sandboxnet (hunk 6 above)
Comment (adaszko) [reply bootstrap.sandboxnet]: This is a separate commit so that I can remove it just before merging.

File: autopropose.py (hunk 8 above)
Comment (adaszko) [reply autopropose.py]: This how the context managers work.  You get an exception from `yield` it it gets raised within the body of `with` statement.

File: autopropose.py (hunk 9 above)
Comment (adaszko) [reply autopropose.py]: Does that result in proposing an empty block if no contract has been deployed?

File: autopropose.py (hunk 7 above)
Comment (adaszko) [reply autopropose.py]: Sure.

File: run-setup (hunk 5 above)
Comment (adaszko) [reply run-setup]: Fair enough.

File: autopropose.py (hunk 9 above)
Comment (zsluedem) [reply autopropose.py]: Why not just exit? Would this make the process to be a zombie?

**Pull request merged**